common/quotas/global/algorithm/requestweighted.go

```diff
@@ -0,0 +1,351 @@
+		var next history
+		prev := ih[id]
+		aps := PerSecond(float64(req.Accepted) / float64(elapsed/time.Second))
+		rps := PerSecond(float64(req.Rejected) / float64(elapsed/time.Second))
+		if prev.lastUpdate.IsZero() {
+			next = history{
+				lastUpdate: snap.now,
+				accepted:   aps, // no history == 100% weight
+				rejected:   rps, // no history == 100% weight
+			}
+		} else {
+			// account for missed updates.
+			// ignoring gc as we are updating, and a very-low reduce just means similar results to zero data.
+			reduce, _ := missedUpdateScalar(snap.now.Sub(prev.lastUpdate), snap)
+
+			next = history{
+				lastUpdate: snap.now,
+				accepted:   weighted(aps, prev.accepted*reduce, snap.weight),
+				rejected:   weighted(rps, prev.rejected*reduce, snap.weight),
+			}
+		}
+
+		ih[id] = next
+		a.usage[key] = ih
+	}
+}
+
+// getWeightsLocked returns the weights of observed hosts (based on ALL requests), and the total number of requests accepted per second.
+func (a *impl) getWeightsLocked(key Limit, snap configSnapshot) (weights map[Identity]HostWeight, usedRPS PerSecond) {
+	ih := a.usage[key]
+	if len(ih) == 0 {
+		return
+	}
+
+	weights = make(map[Identity]HostWeight, len(ih))
+	total := HostWeight(0.0)
+	for id, history := range ih {
+		// account for missed updates
+		reduce, gc := missedUpdateScalar(snap.now.Sub(history.lastUpdate), snap)
+		if gc {
+			// old, clean up
+			delete(ih, id)
+			continue
+		}
+
+		actual := HostWeight((history.accepted + history.rejected) * reduce)
+
+		weights[id] = actual // populate with the reduced values so it doesn't have to be calculated again
+		total += actual      // keep a running total to scale all values when done
+		usedRPS += history.accepted * reduce
+	}
+
+	if len(ih) == 0 {
+		// completely empty Limit, gc it as well
+		delete(a.usage, key)
+		return nil, 0
+	}
+
+	for id := range ih {
+		// scale by the total.
+		// this also ensures all values are between 0 and 1 (inclusive)
+		weights[id] = weights[id] / total
+	}
+	return weights, usedRPS
+}
+
+func (a *impl) HostWeights(host Identity, limits []Limit) (weights map[Limit]HostWeight, usedRPS map[Limit]PerSecond) {
+	a.mut.Lock()
+	defer a.mut.Unlock()
+
+	weights = make(map[Limit]HostWeight, len(limits))
+	usedRPS = make(map[Limit]PerSecond, len(limits))
+	snap := a.snapshot()
+	for _, lim := range limits {
+		hosts, used := a.getWeightsLocked(lim, snap)
+		if len(hosts) > 0 {
+			usedRPS[lim] = used // limit is known, has some usage
+			if weight, ok := hosts[host]; ok {
+				weights[lim] = weight // host has a known weight
+			}
+		}
+	}
+	return weights, usedRPS
+}
+
+func (a *impl) GC() Metrics {
+	a.mut.Lock()
+	defer a.mut.Unlock()
+
+	m := Metrics{}
+	snap := a.snapshot()
+	// TODO: too costly? can check the first-N% each time and it'll eventually visit all keys, demonstrated in tests.
+	for lim, dat := range a.usage {
+		for host, hist := range dat {
+			if _, gc := missedUpdateScalar(snap.now.Sub(hist.lastUpdate), snap); gc {
```

Comment:
results so far look like (while minimizing wasted calcs):
- sadly a bit larger and more "you have to know to check it" rather than having all things you need in one call
- reduces one unnecessary `math.Pow`

... not entirely sure how I feel about that tradeoff, but feels defensible anyway.
I'll push an update with that soon.